In [1]:
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
Turkey = pd.read_excel("ASHTURR5.xlsx")
Austria = pd.read_excel("ASHAUTR5.xlsx")
Germany = pd.read_excel("ASHDEUR5.xlsx")
Egypt = pd.read_excel("ASHEGYR5.xlsx")
France = pd.read_excel("ASHFRAR5.xlsx")
Iran = pd.read_excel("ASHIRNR5.xlsx")
Jordan = pd.read_excel("ASHJORR5.xlsx")
Netherlands = pd.read_excel("ASHNLDR5.xlsx")
Sweden = pd.read_excel("ASHSWER5.xlsx")

HC is the home context data. This is a questionnaire filled in by the parents of each student. It contains crucial information on the immigration status of students and the languages they speak.

In [2]:
Turkey.columns.to_list()

['IDCNTRY',
 'IDPOP',
 'IDGRADER',
 'IDGRADE',
 'WAVE',
 'IDSCHOOL',
 'IDCLASS',
 'IDSTUD',
 'ITLANG_HQ',
 'LCID_HQ',
 'ASBH01A',
 'ASBH01B',
 'ASBH01C',
 'ASBH01D',
 'ASBH01E',
 'ASBH01F',
 'ASBH01G',
 'ASBH01H',
 'ASBH01I',
 'ASBH01J',
 'ASBH01K',
 'ASBH01L',
 'ASBH01M',
 'ASBH01N',
 'ASBH01O',
 'ASBH01P',
 'ASBH01Q',
 'ASBH01R',
 'ASBH02A',
 'ASBH02B',
 'ASBH03A',
 'ASBH03B',
 'ASBH03C',
 'ASBH03D',
 'ASBH03E',
 'ASBH03F',
 'ASBH04',
 'ASBH05AA',
 'ASBH05AB',
 'ASBH05B',
 'ASBH06',
 'ASBH07A',
 'ASBH07B',
 'ASBH07C',
 'ASBH07D',
 'ASBH07E',
 'ASBH07F',
 'ASBH07G',
 'ASBH08A',
 'ASBH08B',
 'ASBH08C',
 'ASBH08D',
 'ASBH08E',
 'ASBH08F',
 'ASBH09',
 'ASBH10',
 'ASBH11A',
 'ASBH11B',
 'ASBH11C',
 'ASBH11D',
 'ASBH11E',
 'ASBH11F',
 'ASBH11G',
 'ASBH11H',
 'ASBH12',
 'ASBH13',
 'ASBH14A',
 'ASBH14B',
 'ASBH14C',
 'ASBH15A',
 'ASBH15B',
 'ASBH16',
 'ASBH17A',
 'ASBH17B',
 'ASBH18AA',
 'ASBH18AB',
 'ASBH18BA',
 'ASBH18BB',
 'ASBH18CA',
 'ASBH18CB',
 'ASBH18DA',
 'ASBH18DB',
 'ASBH18EA',
 '

In [3]:
columns_to_keep = ['IDCNTRY','IDSTUD','ASBH02A','ASBH02B','ASBH03A','ASBH04', 'ASBH15A','ASBH15B','ASBH16','ASBH17A','ASBH17B','ASBH18AA','ASBH18AB']

In [11]:
def select_columns(dataframe, columns_to_keep):
    """
    Selects specified columns from a DataFrame and replaces the original DataFrame with the new one.
    
    Parameters:
        dataframe (pd.DataFrame): The input DataFrame.
        columns_to_keep (list): A list of column names to keep.
        
    Returns:
        None
    """
   # Modify the original DataFrame inplace
    dataframe.drop(dataframe.columns.difference(columns_to_keep), axis=1, inplace=True)

In [5]:
Countries = [Turkey, Austria, Germany, Egypt, France, Iran, Jordan, Netherlands, Sweden]

In [12]:
for country in Countries:
    select_columns(country, columns_to_keep)

In [13]:
Turkey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6032 entries, 0 to 6031
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   IDCNTRY   6032 non-null   int64  
 1   IDSTUD    6032 non-null   int64  
 2   ASBH02A   5624 non-null   float64
 3   ASBH02B   5624 non-null   float64
 4   ASBH03A   5624 non-null   float64
 5   ASBH04    5624 non-null   float64
 6   ASBH15A   5624 non-null   float64
 7   ASBH15B   5624 non-null   float64
 8   ASBH16    5624 non-null   float64
 9   ASBH17A   5624 non-null   float64
 10  ASBH17B   5624 non-null   float64
 11  ASBH18AA  5624 non-null   float64
 12  ASBH18AB  5624 non-null   float64
dtypes: float64(11), int64(2)
memory usage: 612.8 KB


In [14]:
df = pd.concat([Turkey, Austria, Germany, Egypt, France, Iran, Jordan, Netherlands, Sweden])

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50367 entries, 0 to 5174
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   IDCNTRY   50367 non-null  int64 
 1   IDSTUD    50367 non-null  int64 
 2   ASBH02A   41503 non-null  object
 3   ASBH02B   16269 non-null  object
 4   ASBH03A   41179 non-null  object
 5   ASBH04    41231 non-null  object
 6   ASBH15A   37405 non-null  object
 7   ASBH15B   34597 non-null  object
 8   ASBH16    40406 non-null  object
 9   ASBH17A   36503 non-null  object
 10  ASBH17B   32361 non-null  object
 11  ASBH18AA  40161 non-null  object
 12  ASBH18AB  35131 non-null  object
dtypes: int64(2), object(11)
memory usage: 5.4+ MB


In [16]:
df.rename(columns={'IDCNTRY': 'Country', 'ASBH02A':'Immigrant','ASBH02B':'AgeArrival','ASBH03A':'CountryLanguageBeforeSchool','ASBH04':'CountryLanguageHome','ASBH15A':'EducationParentA','ASBH15B':'EducationParentB','ASBH16':'EducationHope','ASBH17A':'ParentAJob','ASBH17B':'ParentBJob','ASBH18AA':'ParentACountryLanguage','ASBH18AB':'ParentBCountryLanguage'}, inplace=True)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50367 entries, 0 to 5174
Data columns (total 13 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Country                      50367 non-null  int64 
 1   IDSTUD                       50367 non-null  int64 
 2   Immigrant                    41503 non-null  object
 3   AgeArrival                   16269 non-null  object
 4   CountryLanguageBeforeSchool  41179 non-null  object
 5   CountryLanguageHome          41231 non-null  object
 6   EducationParentA             37405 non-null  object
 7   EducationParentB             34597 non-null  object
 8   EducationHope                40406 non-null  object
 9   ParentAJob                   36503 non-null  object
 10  ParentBJob                   32361 non-null  object
 11  ParentACountryLanguage       40161 non-null  object
 12  ParentBCountryLanguage       35131 non-null  object
dtypes: int64(2), object(11)
memory u

ASBH02A	GEN\CHILD BORN IN <COUNTRY>	HQ-02A	Nominal	1	0			 1: Yes; 2: No	9: Omitted or invalid; Sysmis: Not administered

ASBH02B	GEN\AGE OF CHILD WHEN CAME TO <COUNTRY>	HQ-02B	Ordinal	1	0			 1: Younger than 3 years old; 2: 3 to 5 years old; 3: 6 to 7 years old; 4: 8 years old or older	9: Omitted or invalid; 6: Logically not applicable; Sysmis: Not administered

ASBH03A	GEN\LANGUAGE\<LANGUAGE OF TEST>	HQ-03a	Nominal	1	0			 1: Yes; 2: No	9: Omitted or invalid; Sysmis: Not administered

ASBH04	GEN\HOW OFTEN <LANG OF TEST> AT HOME	HQ-04	Ordinal	1	0			 1: Always; 2: Almost always; 3: Sometimes; 4: Never	9: Omitted or invalid; Sysmis: Not administered

ASBH15A	GEN\LVL OF EDUCATION\GUARDIAN A	HQ-15a	Ordinal	2	0			 1: Did not go to school; 10: Not applicable; 2: Some <Primary education—ISCED Level 1 or Lower secondary education—ISCED Level 2>; 3: <Lower secondary education—ISCED Level 2>; 4: <Upper secondary education—ISCED Level 3>; 5: <Post-secondary, non-tertiary education— ISCED Level 4>; 6: <Short-cycle tertiary education—ISCED Level 5>; 7: <Bachelor’s or equivalent level—ISCED Level 6>; 8: <Master’s or equivalent level—ISCED Level 7>; 9: <Doctor or equivalent level—ISCED Level 8>	999: Omitted or invalid; Sysmis: Not administered

ASBH15B	GEN\LVL OF EDUCATION\GUARDIAN B	HQ-15b	Ordinal	2	0			 1: Did not go to school; 10: Not applicable; 2: Some <Primary education—ISCED Level 1 or Lower secondary education—ISCED Level 2>; 3: <Lower secondary education—ISCED Level 2>; 4: <Upper secondary education—ISCED Level 3>; 5: <Post-secondary, non-tertiary education— ISCED Level 4>; 6: <Short-cycle tertiary education—ISCED Level 5>; 7: <Bachelor’s or equivalent level—ISCED Level 6>; 8: <Master’s or equivalent level—ISCED Level 7>; 9: <Doctor or equivalent level—ISCED Level 8>	999: Omitted or invalid; Sysmis: Not administered

Gamal was interested in comparing the treatment of kids in different countries - he thought treatment in Egypt would be poor. So that's something we can easily investigate! 

# So next steps - we want to replace the country ID with the name of the country. Should be easy enough. We're gonna do some groupbys and some pivot tables. 

I think what I want to do is read in all the student context stuff and then we can start doing some nice comparisons. 

SC is the Student Context data which is a questionnaire filled in by the students themselves as well as some totals on their achievement in the test. As we can see there are 6032 children who took part in this assessment in Turkey.

In [55]:
TurkeySC = pd.read_excel("ASGTURR5.xlsx")
AustriaSC = pd.read_excel("ASGAUTR5.xlsx")
EgyptSC = pd.read_excel("ASGEGYR5.xlsx")
FranceSC = pd.read_excel("ASGFRAR5.xlsx")
IranSC = pd.read_excel("ASGIRNR5.xlsx")
JordanSC = pd.read_excel("ASGJORR5.xlsx")
NetherlandsSC = pd.read_excel("ASGNLDR5.xlsx")
SwedenSC = pd.read_excel("ASGSWER5.xlsx")
GermanySC = pd.read_excel("ASGDEUR5.xlsx")

In [56]:
SCdf = pd.concat([TurkeySC, AustriaSC, EgyptSC, FranceSC, IranSC, JordanSC, NetherlandsSC, SwedenSC, GermanySC])

In [57]:
SCdf.columns.to_list()

['IDCNTRY',
 'IDPOP',
 'IDGRADER',
 'IDGRADE',
 'WAVE',
 'IDSCHOOL',
 'IDCLASS',
 'IDSTUD',
 'ITSEX',
 'ITADMINI',
 'LCID_SA',
 'LCID_SQ',
 'ITLANG_SA',
 'ITLANG_SQ',
 'IDBOOK',
 'ASBG01',
 'ASBG03',
 'ASBG04',
 'ASBG05A',
 'ASBG05B',
 'ASBG05C',
 'ASBG05D',
 'ASBG05E',
 'ASBG05F',
 'ASBG05G',
 'ASBG05H',
 'ASBG05I',
 'ASBG05J',
 'ASBG05K',
 'ASBG06',
 'ASBG07A',
 'ASBG07B',
 'ASBG08A',
 'ASBG08B',
 'ASBG09A',
 'ASBG09B',
 'ASBG09C',
 'ASBG09D',
 'ASBG09E',
 'ASBG09F',
 'ASBG09G',
 'ASBG09H',
 'ASBG10A',
 'ASBG10B',
 'ASBG10C',
 'ASBG10D',
 'ASBG10E',
 'ASBG10F',
 'ASBG11A',
 'ASBG11B',
 'ASBG11C',
 'ASBG11D',
 'ASBG11E',
 'ASBG11F',
 'ASBG11G',
 'ASBG11H',
 'ASBG11I',
 'ASBG11J',
 'ASBR01A',
 'ASBR01B',
 'ASBR01C',
 'ASBR01D',
 'ASBR01E',
 'ASBR01F',
 'ASBR01G',
 'ASBR01H',
 'ASBR01I',
 'ASBR02A',
 'ASBR02B',
 'ASBR02C',
 'ASBR02D',
 'ASBR02E',
 'ASBR03A',
 'ASBR03B',
 'ASBR03C',
 'ASBR04',
 'ASBR05',
 'ASBR06A',
 'ASBR06B',
 'ASBR07A',
 'ASBR07B',
 'ASBR07C',
 'ASBR07D',
 'ASBR07E',


In [58]:
SCdf = SCdf[['IDCNTRY',
 'IDSTUD','ASBG01',
 'ASBG03','ASBG10C',
 'ASBG10D',
 'ASBG10E',
 'ASBG10F',
 'ASBG11A',
 'ASBG11B',
 'ASBG11C',
 'ASBG11D',
 'ASBG11E',
 'ASBG11F',
 'ASBG11G',
 'ASBG11H',
 'ASBG11I',
 'ASBG11J','ASDAGE','ASRREA01',
 'ASRREA02',
 'ASRREA03',
 'ASRREA04',
 'ASRREA05',
 'ASRLIT01',
 'ASRLIT02',
 'ASRLIT03',
 'ASRLIT04',
 'ASRLIT05',
 'ASRINF01',
 'ASRINF02',
 'ASRINF03',
 'ASRINF04',
 'ASRINF05',
 'ASRIIE01',
 'ASRIIE02',
 'ASRIIE03',
 'ASRIIE04',
 'ASRIIE05',
 'ASRRSI01',
 'ASRRSI02',
 'ASRRSI03',
 'ASRRSI04',
 'ASRRSI05']]

In [60]:
SCdf= SCdf.rename(columns={'IDCNTRY': 'Country'})

In [61]:
DF = pd.merge(df, SCdf, on=None)

In [63]:
DF["CountryLanguageBeforeSchool"].value_counts()

Yes    25915
1.0    12270
2.0     1477
No      1175
9.0      342
Name: CountryLanguageBeforeSchool, dtype: int64

In [64]:
def populate_column(row):
    if row['AgeArrival'] == 1:
        return 0
    elif row['AgeArrival'] == 2:
        return 3
    elif row['AgeArrival'] == 3:
        return 6
    elif row['AgeArrival'] == 4:
        return 8
    else:
        return None  # Or any default value

# Apply the function to each row
DF['MINAGEARRIVAL'] = DF.apply(populate_column, axis=1)

In [65]:
DF['MINAGEARRIVAL'].value_counts()

0.0    182
3.0    160
6.0    108
8.0     63
Name: MINAGEARRIVAL, dtype: int64

In [66]:
DF['rounded_ages'] = DF['ASDAGE'].round()

In [67]:
DF['rounded_ages'].value_counts()

10.0    38410
11.0     9104
9.0      1964
12.0      758
13.0       83
8.0        18
14.0       17
99.0        3
7.0         1
Name: rounded_ages, dtype: int64

In [68]:
mean_of_belonging = DF.groupby('Immigrant').agg({'ASBG10C': 'mean'})
print(mean_of_belonging)

TypeError: unsupported operand type(s) for +: 'float' and 'str'

In [69]:
DF['Immigrant'].value_counts()

Yes    25072
1.0    13403
No      2342
2.0      548
9.0      138
Name: Immigrant, dtype: int64

In [70]:
bool = DF["Immigrant"] == 'No'

First thing I need to do is recognise these countries and add these values to that column. 

In [78]:
# Mapping dictionary
update_dict = {40: 'Austria', 818: 'Egypt', 250:'France', 276:'Germany',364:'Iran',400:'Jordan',528:'Netherlands',752:'Sweden',792:'Turkey'}

# Updating the column using map
DF['Country'] = DF['Country'].map(update_dict).fillna(DF['Country'])

In [79]:
DF['Country'].value_counts()

Egypt          7979
Jordan         6150
Turkey         6032
Iran           5962
France         5339
Sweden         5175
Austria        4806
Germany        4611
Netherlands    4313
Name: Country, dtype: int64

In [80]:
# Mapping dictionary
update_dict2 = {1.0: 'Yes',2.0:'No' }

# Updating the column using map
DF['Immigrant'] = DF['Immigrant'].map(update_dict2).fillna(DF['Immigrant'])
DF['Immigrant'].value_counts()

Yes    38475
No      2890
9.0      138
Name: Immigrant, dtype: int64

In [82]:
# Mapping dictionary
update_dict3 = { 'Yes':'No','No':'Yes' }

# Updating the column using map
DF['Immigrant'] = DF['Immigrant'].map(update_dict3).fillna(DF['Immigrant'])
DF['Immigrant'].value_counts()

No     38475
Yes     2890
9.0      138
Name: Immigrant, dtype: int64

So we have 2890 children who were not born in the country where they took the test.

In [83]:
DF["CountryLanguageBeforeSchool"].value_counts()

Yes    25915
1.0    12270
2.0     1477
No      1175
9.0      342
Name: CountryLanguageBeforeSchool, dtype: int64

In [84]:
# Mapping dictionary
update_dict4 = {1.0: 'Yes',2.0:'No' }

# Updating the column using map
DF['CountryLanguageBeforeSchool'] = DF['CountryLanguageBeforeSchool'].map(update_dict2).fillna(DF['CountryLanguageBeforeSchool'])
DF['CountryLanguageBeforeSchool'].value_counts()

Yes    38185
No      2652
9.0      342
Name: CountryLanguageBeforeSchool, dtype: int64

And we have 2652 children who did not speak the language of the test before they started school. So some children were born elsewhere but were already speaking the language of the test before they started school.

I think I need to tidy this up a bit and probably want to put some of it on github.

In [ ]:
def rename_columns_for_presentation(df):
    column_mapping = {
        'col1': 'User_ID',
        'col2': 'Transaction_Date',
        'col3': 'Amount'
    }
    return df.rename(columns=column_mapping)